File: 1675.minimize-deviation-in-array.cs
/*
 * @lc app=leetcode id=1675 lang=csharp
 *
 * [1675] Minimize Deviation in Array
 */

// @lc code=start
using System.Collections.Generic;
public class Solution 
{
    public int[] SmallestRange(IList<IList<int>> nums) 
    {
        int globalMax = 100000;
        int globalMin = - 100000;
        int maxSoFar = - 100000;  
        PriorityQueue<int[], int> pq = new PriorityQueue<int[], int>();

        for (int i = 0; i <  nums.Count; i++)
        {
            globalMax = Math.Max(globalMax, nums[i][nums[i].Count - 1]);
            globalMin = Math.Min(globalMin, nums[i][0]);
            pq.Enqueue(new int[2]{i, 0}, nums[i][0]);
            maxSoFar = Math.Max(maxSoFar, nums[i][0]);
        }   

        int[] result = new int[2]{globalMin, globalMax};
        int diff = globalMax - globalMin;

        while(pq.TryDequeue(out int[] node, out int num))
        {
            if (maxSoFar - num < diff)
            {
                diff = maxSoFar - num;
                result[0] = num;
                result[1] = maxSoFar;
            }
            if (node[1] < nums[node[0]].Count - 1)
            {
                pq.Enqueue(new int[2]{node[0], node[1] + 1}, nums[node[0]][node[1] + 1]);
                maxSoFar = Math.Max(maxSoFar, nums[node[0]][node[1] + 1]);
            }
            else
            {
                break;
            }
        }
        return result;
    }
    public int MinimumDeviation(int[] nums) 
    {
        List<IList<int>> data = new List<IList<int>>();
        foreach (int num in nums)
        {
            List<int> datum = new List<int>();
            if (num % 2 == 1)
            {
                datum.Add(num);
                datum.Add(2 * num);
            }
            else
            {
                int temp = num;
                while (temp % 2 == 0)
                {
                    datum.Add(temp);
                    temp /= 2;
                }
                datum.Add(temp);
                datum.Reverse();
            }
            data.Add(datum);
        }

        int[] smallestR = SmallestRange(data);
        return smallestR[1] - smallestR[0];
    }
}
// @lc code=end

File: 1675.minimize-deviation-in-array.py
#
# @lc app=leetcode id=1675 lang=python3
#
# [1675] Minimize Deviation in Array
#

# @lc code=start
import heapq
from typing import List


class Solution:
    '''
    This is the same problem as 632.
    If num is odd, it can only be [num, 2*num]
    If num is even, it can be [num // (2^n), ... ,num // 2, num] in which num // (2^n) is odd
    '''
    def smallestRange(self, nums: List[List[int]]) -> List[int]:
        globalMax, globalMin, maxSoFar = - float('inf'), float('inf'), - float('inf')
        heap = []
        for i, num in enumerate(nums):
            globalMax = max(globalMax, num[-1])
            globalMin = min(globalMin, num[0])
            heapq.heappush(heap, [num[0], i, 0])
            maxSoFar = max(maxSoFar,num[0])
            
        result = [globalMin, globalMax]
        diff = globalMax - globalMin
        
        while heap:
            num, row, col = heapq.heappop(heap)
            if maxSoFar - num < diff:
                diff = maxSoFar - num
                result = [num, maxSoFar]
            if col < len(nums[row]) - 1:
                heapq.heappush(heap, [nums[row][col + 1], row, col + 1])
                maxSoFar = max(maxSoFar, nums[row][col + 1])
            else:
                break
        return result


    def minimumDeviation(self, nums: List[int]) -> int:
        data = []
        for num in nums:
            datum = []
            if num % 2 == 1:
                datum.append(num)
                datum.append(2 * num)
            else:
                temp = num
                while temp % 2 == 0:
                    datum.append(temp)
                    temp //= 2
                datum.append(temp)
                datum.reverse()
            data.append(datum)
        
        smallestR = self.smallestRange(data)
        return smallestR[1] - smallestR[0] 
        
# @lc code=end

